plaidml/bridge/openvino/ops/psroi_pooling.cpp

```diff
@@ -0,0 +1,256 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "plaidml_ops.hpp"
+#include "plaidml_util.hpp"
+
+#include "ngraph/opsets/opset.hpp"
+#include "ngraph/opsets/opset5.hpp"
+
+#include "plaidml/op/op.h"
+
+using namespace plaidml;          // NOLINT[build/namespaces]
+using namespace InferenceEngine;  // NOLINT[build/namespaces]
+using namespace edsl;             // NOLINT[build/namespaces]
+
+namespace {
+
+template <typename T>
+std::vector<T> cast_constant_operand(size_t operand_idx, ngraph::Node* layer) {
+  auto* ngraph_const = ngraph::as_type<ngraph::op::Constant>(layer->get_input_node_ptr(operand_idx));
+  if (ngraph_const) {
+    return ngraph_const->cast_vector<T>();
+  } else {
+    THROW_IE_EXCEPTION
+        << "Dynamic coordinate is not currently supported by PlaidML plugin; all coordinate must be Constants. ";
+  }
+}
+
+// Get single bin output
+edsl::Tensor single_bin_pooling(edsl::Tensor I, size_t ph, size_t pw, size_t group_size) {
+  std::vector<edsl::TensorDim> I_dims(4), O_dims(4);
+  std::vector<edsl::TensorIndex> I_idxs(4), O_idxs(4);
+  I.bind_dims(I_dims);
+  O_dims[0] = O_dims[2] = O_dims[3] = edsl::TensorDim(1);
+  O_dims[1] = I_dims[1] / (group_size * group_size);
+  I_idxs[1] = O_idxs[1] * group_size * group_size + ph * group_size + pw;
+  edsl::Tensor O = (Contraction(O_dims, O_idxs).sum(I(I_idxs))) / (I_dims[2] * I_dims[3]);
+  return O;
+}
+
+edsl::Tensor extract_reduce_rois_dim(edsl::Tensor I) {
```

Comment:
Those extract functions seems to indicate we are generating a lot of redundant data, and you need these functions to collect a subset of the data. Please check if we can avoid generating the redundant data.

Reply:
If the results of gather_ND can match the case in PSROIPooling when index is out of bound, that can be replaced by gather_ND and bilinear interpolation. But i don't think it can work by using gather_ND, because that it isn't implemented to use gather_ND and bilinear interpolation to match the results when index is out of bound.

Reply:
OK, Let's aim to submit a draft PR to plaidml by the end of tomorrow. I think there is a lot of items to discuss for this work.